### ISSUE TITLE
myTimer.elapsed()

### ISSUE THREAD

**Issue by vangalvin** [thomasfredericks/Chrono#4]
Thank you for such an easy to use timing system.
The only issue I have seen so far is that the myTimer.elapsed() returns the results of all of only the first times that was created and seems to not be limited to just the instance of the intended timer

I have 4 timers running
Chrono t1(Chrono::SECONDS);
Chrono t2(Chrono::SECONDS);
Chrono t3(Chrono::SECONDS);
Chrono t4(Chrono::SECONDS);

each one is set for a different time however when I check the times in their respective loops I only get the elapsed time of the first timer unless i force a restart on that timer before it is started.

**Comment by vangalvin** [comment]
What I am trying to do is keep a record of how long a the specific function took to run so i can use the time taken in another function. 

Essentially I am activating a water valve, there is a sensor on the tank that tells me when the tank is full, once its full then i can use that time to estimate how long the tank is going to take to empty. Start the pump and time it using the time taken to fill it as there is no sensor that tells me when its empty.

**Comment by sofian** [comment]
Hi. Since you are using seconds, it is possible that if you start all Chrono objects at the same time, they will all return the same elapsed() value. Could you send us your full code?

**Comment by sofian** [comment]
Chrono objects are automatically started at creation. This is what confuses you. You need to call start() on the object when you want to reset its timer.

**Comment by sofian** [comment]
It's possible not to start automatically by declaring Chrono chrono(Chrono::SECONDS, false).

**Comment by vangalvin** [comment]
awesome, thank you sofian :)
I managed to get that sorted and putting a reset in also gives a bit of a safety factor now. 
Will post some of the code when I am done, the controller on the washing machine blew up and a replacement was going to cost an arm and a leg so I decided to use an arduino :)
using the Chrono extensively to handle all of the functions such as WaterFill, agitate cycles, spin cycles etc..

**Comment by sofian** [comment]
I'm thinking that the fact that chronos automatically start should be clarified in the README file. Furthermore, since the "custom function" constructor (Chrono(chrono_t (*getTime_)(void), bool startNow=true)) has an option "startNow" that allows to prevent autostart, it would be more sound/systematic to offer that option "startNow" in other constructors.

@thomasfredericks I made a change to the code in that respect, please have a look and let me know what you think (it's in a "develop" branch so it's not affecting the master branch):
https://github.com/thomasfredericks/Chrono/commit/6b5e6af8d6cd7365181cc2d532396b2056ae8d65

**Comment by vangalvin** [comment]
I will fire it in later tonight, foun another issue with the hardware that is not helping things.

Have the nano driving an 8 channel relay board and turning the motor on and off.  This seems to be creating some rfi that is feeding back to the nano and making the clock unstable.

Going to make some snubbers up tonight and see if that fixes it as we'll as house the nano and relays in seperate metal boxes

Sent from my iPhone

On 4/03/2017, at 2:02 pm, "Sofian Audry" <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi. Since you are using seconds, it is possible that if you start all Chrono objects at the same time, they will all return the same elapsed() value. Could you send us your full code?

**Comment by thomasfredericks** [comment]
For me, the simple basic Chrono is just enough. If you want to add more features @sofian go ahead, as long as the features are documented in the documentation (not the Readme) and a new release is made so it syncs with the Arduino IDE. But the way, the documentation will be converted to a GitHub page eventually.

**Comment by sofian** [comment]
@vangalvin If you are using interrupts, it might impact (slow down) the micros() and millis() function which will in turn impact the Chrono objects. An alternative is to use a RTC (real time clock). There is an easy way to use a RTC in conjunction with Chrono.

**Comment by vangalvin** [comment]
Please keep in mind that the code is not clean and there are some things that still need to be sorted out but it works.. well works with the relay board just not when the motor is connected. :(


`#include <Chrono.h>
#include <Bounce2.h>
// Arduion Pin Allocations
// Constants for Control Pins
const int LedPin =  13;         // LED Pin
                                // Washingmachine Sensor Pins
const int WaterLevelPin = 2;    // N/C Switch to sense the water level If it goes open then stop everyting.
const int LindSwitchPin = 3;    // Goes OPEN circut if the lit is opend. E-STOP
const int SpinSensorPin = 12;   // When the spin locking pin is in place this is short
                                // IN# are the relay inputs for the Duinotech 8 Channel Board
const int CWPin = 4;            // IN1 - Clockwise Pin
const int CCWPin = 5;           // IN2 - Counter Clockwise / FAST SPIN Pin.
const int SpinActuatorPin = 6;  // IN3 - Pull the locking pin in for the spin cycle
const int ColdWaterValve = 7;   // IN4 - Cold Water Valve
const int HotWaterValve = 8;    // IN5 - Cold Water Valve
const int WaterPumpPin = 9;     // IN6 - Water Pump
                                // Control Pannel Pins
const int StartPin = 10;        // Restart After and Error
const int RestartPin = 11;      // Restart After and Error

int fill_time = 0;
unsigned long agitate_time = 0;
unsigned long agitate_loop_time = 0;
unsigned long balance_time = 0;
unsigned long balance_loop_time = 0;
unsigned long spin_time = 0;
unsigned long empty_time = 0;

// Control Variables
int ledState = LOW;             // ledState used to set the LED
int CWState = LOW;              // Clockwise Relay State
int CCWState = LOW;             // Counter Clockwise Relay State
int SpinState = LOW;            // Spint Actuator State
int ColdState = LOW;            // Cold Water Valve State
int HotState = LOW;             // Hot Water Valve State
int PumpState = LOW;            // Water Pump State

// Washing Cycle Times
int fill_v = 10;
int agitate_v = 120; //180;
long agitate_loop_v = 1;
int balance_v = 200;
int balance_loop = 0;
int balance_loop_v1 = 5;
int balance_loop_v2 = 2;
int spin_v = 200;
int shortSpin_v = 10;
int rinseSpin_v = 200;
int rinseSpinWater_v = 6;
int minFillTime_v = 400; //368;
int maxFillTime_v = 600; // 621;
int minEmptyTime_v = 250;
int maxEmptyTime_v = 350;
int soakTime_v = 60;
int spinBrakeDelay = 40;

// Process and Cycle Variables
int MachineCycle = 1;
int MachineProcess = 0;
int WashingCycles = 0;          // this is where we hold the variable as to how many times we cycle
int eStop = 0;                  //emergencey stop should be set up on an interupt for the lid
int incomingByte = 0;           // for incoming serial data
int agitateEvent = 0;           // Agitation Event to swith from CW to CCW
int spinEnable = 0;             // enable spin

String lastmessage;

Chrono fillTimer(Chrono::SECONDS);
Chrono agitateTimer(Chrono::SECONDS);
Chrono agitateLoopTimer(Chrono::SECONDS);
Chrono balanceTimer(Chrono::SECONDS);
Chrono balanceLoopTimer(Chrono::SECONDS);
Chrono spinTimer(Chrono::SECONDS);
Chrono rinseSpinSpin(Chrono::SECONDS);
Chrono rinseSpinWater(Chrono::SECONDS);
Chrono waterFillTime(Chrono::SECONDS);
Chrono waterEmptyTime(Chrono::SECONDS);
Chrono soakTime(Chrono::SECONDS);
Chrono spinBrakeTime(Chrono::SECONDS);

// Instantiate a Bounce object
Bounce debouncer = Bounce();

void setup() {
  Serial.begin(9600);                   // Serial Communications for Debuging
  pinMode(StartPin,INPUT_PULLUP);       // Pin to use for the Start Button
  debouncer.attach(StartPin);           // After setting up the button, setup the Bounce instance :
  debouncer.interval(40);                // interval in ms
  
 // set the digital pin as output:
  pinMode(LedPin, OUTPUT);              // Onboard LED on pin 13
  pinMode(WaterLevelPin, INPUT_PULLUP); // Enable Pullup for WaterLevel Sensor
  pinMode(LindSwitchPin, INPUT_PULLUP); // Enable Pullup for Lid Switch
  
  pinMode(CWPin, OUTPUT);               // Clockwise Motor Control
  pinMode(CCWPin, OUTPUT);              // Counter Closkwise Motor Control
  pinMode(SpinActuatorPin, OUTPUT);     // Spin Actuator Solinoid
  pinMode(ColdWaterValve, OUTPUT);      // Cold Water Valve
  pinMode(HotWaterValve, OUTPUT);       // Hot Water Valve
  pinMode(WaterPumpPin, OUTPUT);        // Water Pump

  pinMode(RestartPin, INPUT_PULLUP);    // Restart/Resume after Error
  pinMode(SpinSensorPin, INPUT_PULLUP); // Enable Pullup for Spin Sensor

  //Reset all timer's
//fillTimer.restart();
fillTimer.stop();
//agitateTimer.restart();
agitateTimer.stop();
//agitateLoopTimer.restart();
agitateLoopTimer.stop();
//balanceTimer.restart();
balanceTimer.stop();
//balanceLoopTimer.restart();
balanceLoopTimer.stop();
//spinTimer.restart();
spinTimer.stop();
//rinseSpinSpin.restart();
rinseSpinSpin.stop();
//rinseSpinWater.restart();
rinseSpinWater.stop();
//waterFillTime.restart();
waterFillTime.stop();
//waterEmptyTime.restart();
waterEmptyTime.stop();
//soakTime.restart();
soakTime.stop();
spinBrakeTime.stop();
}

void loop(){
  
  debouncer.update();                   // Update the Bounce instance :
  int value = debouncer.read();         // Get the updated value :
  if ( value == LOW ) {
    Serial.println("Washing Cycle Started");
    MachineCycle = 1;                    // If the button is pushed start the machine
    MachineProcess = 1;   
  }
  
   // send data only when you receive data:
  if (Serial.available() > 0) {
    incomingByte = Serial.read();       // read the incoming byte:
    Serial.print("I received: ");       // say what you got:
    Serial.println(incomingByte, DEC);
  }

// If we receive a 0 on the serial interface start the machine
if(incomingByte == 48){
  ColdWaterOff();
  Serial.println("Stop Fill");
  MachineProcess = 0;
}

// If we receive a 1 run this
if(incomingByte == 49){
  Serial.println("Fill with Water");
  MachineProcess = 1;
}

// If we receive a 2 run this
if(incomingByte == 50){
  Serial.println("Agitate Washing");
  MachineProcess = 2;
}

// If we receive a 3 Ballance Load
if(incomingByte == 51){
  Serial.println("Soak Washing");
  MachineProcess = 3;
}

// If we receive a 4 Full Spin
if(incomingByte == 52){
  Serial.println("Agitate Washing");
  MachineProcess = 4;
}

// If we receive a 5 Rinse Spin
if(incomingByte == 53){
  Serial.println("Ballance Spin");
  MachineProcess = 5;
}

// If we receive a 6 run this
if(incomingByte == 54){
  Serial.println("Full Spin");
  MachineProcess = 6;
}

// If we receive a 7 run this
if(incomingByte == 55){
  Serial.println("Full Spin");
  MachineProcess = 7;
}

// If we receive a 8 run this
if(incomingByte == 56){
  Serial.println("Fill with Water");
  MachineProcess = 8;
}

// If we receive a Q run this
if(incomingByte == 113){
  Serial.println("Reset Machine");
  MachineProcess = 13;
}
// If we receive a 9 on the serial interface
if(incomingByte == 57){
  ColdWaterOff(); // turn water off
  pumpOff(); // turn pump off
  StopAgitate(); //
  endSpinCycle(); // 
  estop();
  Serial.println("PRINT STATUS");
  Serial.print("Last Fill Time : ");
  Serial.print(fill_time);
  Serial.print(" Last Empty Time : ");
  Serial.println(empty_time);
  MachineProcess = 0;
}

  //Start Cycle
  if(MachineCycle == 1){

  //Start Process 1 Fill the washing machine with water
      if(MachineProcess == 1){
                
        if(fillWater() != 1){
          printOnce("Filling Washing Machine");
//          Serial.println(fill_time);
        }else{
//          Serial.print("Time taken to fill machine was : ");
//          Serial.println(fill_time);
          MachineProcess = 2; //Move Agitate Process
        }//fill machine 1
      }//End Process 0

  //Start Process 2 Agitate the Washing
      if(MachineProcess == 2){                
        if(agitate() != 1){
          printOnce("Agitate 1");
        }else{
          StopAgitate(); //
          MachineProcess = 3; //Move to next process
        }//fill machine 1
      }//End Process 0

   //Start Process 3 soak wash
      if(MachineProcess == 3){   
        if(soakWash() != 1){
          printOnce("Soak Washing");            
        }else{
          //agitate again
           MachineProcess = 4; //Move to Short Agitate
        }
      }//End soak wash 3

  //Start Process 4 Agitate for a short amount of time
      if(MachineProcess == 4){                
        if(agitate() != 1){
          printOnce("Agitate 2");
        }else{
          StopAgitate(); //
          MachineProcess = 5; //Ballance Spin and Empty
        }//Short Agitate 4
      }//End Process 4
  
  //Start Process 5 Ballance the Load
      if(MachineProcess == 5){
        if(ballanceLoad() != 1){               
          if(emptyMachine() != 1){
            printOnce("Ballance the Load for Spin Cycle");
             //emptyMachine();
          }
        }else{
          MachineProcess = 6;
        } //Ballance Load and Empty
      }//End Process 5

  //Start Process 6 FULL SPIN
      if(MachineProcess == 6){ 
        if(fullSpin() != 1){
          printOnce("Full Speed Spinning");
          pumpOn();
        }else{
          coastSpin(); //
          pumpOff();
          MachineProcess = 7; //Rinse Spin
        }//FULL SPIN               
      }//End Process 6

  //Start Process 7 RINSE SPIN
      if(MachineProcess == 7){ 
        if(rinseSpin() != 1){
          printOnce("Rinse and Spin");
          intSpray();
          pumpOn();
        }else{
//          Serial.println("Rinse Spin Stoped");
          ColdWaterOff();
          coastSpin(); //
          pumpOff();
          MachineProcess = 8; //RINSE SPIN
        }//RINSE SPIN 7               
      }//End Process 0

   //Start Process 8 SPIN
      if(MachineProcess == 8){ 
        if(fullSpin() != 1){
          printOnce("Final Spin");
          pumpOn();
        }else{
          ColdWaterOff();
          coastSpin(); //
          pumpOff();
          //MachineProcess = 0; //Move to next process
          //WashingCycles ++;
          if(WashingCycles == 0){
            Serial.println("First cycle finished, Start Secont cycle");
                  WashingCycles = 1;
                  MachineProcess = 1;
                  MachineCycle = 1;
           }else{
                  Serial.println("Secont cycle finished, Reset System");
                  WashingCycles = 0;
                  MachineProcess = 0;
                  MachineCycle = 0;
                  estop(); //resets the whole system
              }
//         }     estop();           
        }//FINAL SPIN               
      }//End Process 8     
      
  //Start Process 8 Fill the washing machine with water
//      if(MachineProcess == 9){                
//        if(emptyMachine() != 1){
//          Serial.println("empty machine : ");
//          Serial.println(empty_time);
//        }else{
//          Serial.print("Time taken to fill machine was : ");
//          Serial.println(empty_time);
//          MachineProcess = 0; //Move to next process
//        }//fill machine 1
//      }//End Process 9

        //Start Process 8 Fill the washing machine with water
      if(MachineProcess == 13){
        estop();              
          MachineProcess = 0; //Move to next process
          WashingCycles = 0;
      }//End Process 13

  }//End Cycle
}//loop

int soakWash(){
  int resultcode = 0;
//REset and Start the timer is its not running
if(!soakTime.isRunning()){
    soakTime.restart();
  }//reset the timer before we start

 if(soakTime.hasPassed(soakTime_v)){
    resultcode = 1;
    soakTime.restart();
    soakTime.stop();
//    Serial.println("Soak Time Finished");
  }else{
    resultcode = 0;
  }
return resultcode;
}//soakWash

void agitateStop(){
      CCWState = LOW;
      CWState = LOW;
      ledState = LOW;
      SpinState = LOW;
}

void agitateCW(){
      CWState = HIGH;
      CCWState = LOW;
      ledState = HIGH;
      SpinState = LOW;
}
void agitateCCW(){
      CWState = LOW;
      CCWState = HIGH;
      ledState = HIGH;
      SpinState = LOW;
}

int agitate(){
  int resultcode = 0;
 
  if(!agitateTimer.isRunning()){
//    Serial.println("Restarting Timer");
    agitateTimer.restart();
    agitateLoopTimer.restart();
  }//reset the timer before we start

    if(!agitateTimer.isRunning()){
//    Serial.println("Restarting Timer");
    agitateTimer.restart();
    agitateLoopTimer.restart();
  }//reset the timer before we start
  if(agitateTimer.hasPassed(agitate_v)){
    agitateStop();
    resultcode = 1;
    agitateTimer.restart();
    agitateTimer.stop();
//    Serial.println("Agitation Halted");
  }else{
    //Agitate run goes in here
//    Serial.println("Agitation Running");
    if(agitateLoopTimer.hasPassed(agitate_loop_v, true)){
        if(agitateEvent == 1){
            agitateCW();
//            Serial.println("CW ");
        }
        if(agitateEvent == 2){
            agitateStop();
//            Serial.println("STOP ");
        }
        if(agitateEvent == 3){
            agitateCCW();
//            Serial.println("CW ");
        }
        if(agitateEvent == 4){
            agitateStop();
            agitateEvent = 0;   //Loop agitation Event
//            Serial.println("STOP ");
        }
        agitateEvent++;
        resultcode = 0;
    }
}
digitalWrite(CWPin, CWState);
digitalWrite(CCWPin, CCWState);
digitalWrite(LedPin, ledState);
digitalWrite(SpinActuatorPin, SpinState);
return resultcode;
}

void StopAgitate(){
      agitateStop();
      digitalWrite(CWPin, CWState);
      digitalWrite(CCWPin, CCWState);
      digitalWrite(LedPin, ledState);
      digitalWrite(SpinActuatorPin, SpinState);
}

int printOnce(String message){
  if(message != lastmessage){
    lastmessage = message;
    Serial.println(message);
  }
}
void spinUp(){
  CWState = HIGH;
  ledState = HIGH;
  SpinState = HIGH;
}
void spinDown(){
  CWState = LOW;
  ledState = LOW;
  SpinState = HIGH;
}

void stopSpin(){
  CWState = LOW;
  ledState = LOW;
  SpinState = LOW;
}

int spinBrake(){
  int resultcode = 0;

  if(!spinBrakeTime.isRunning()){
    spinBrakeTime.restart();
  }//reset the timer before we stat
  
    if(spinBrakeTime.hasPassed(spinBrakeDelay)){
      stopSpin();
      spinBrakeTime.stop();
//      Serial.println("Brake Applied");
      resultcode = 1;
    }else{
      resultcode = 0;
      spinDown();
//      Serial.println("Waiting for Brake");
    }
    
digitalWrite(LedPin, ledState);
digitalWrite(SpinActuatorPin, SpinState);
digitalWrite(CWPin, CWState);
return resultcode;
}

int fullSpin(){
  SpinState = HIGH;
  int resultcode = 0;

//if(fill_time < minFillTime_v){
//    spin_v = minEmptyTime_v;
//spin_v = maxEmptyTime_v / 2;
//}else{
//spin_v = maxEmptyTime_v / 2;
//}
  
if(digitalRead(SpinSensorPin) == 0){
  
  if(!spinTimer.isRunning()){
    spinTimer.restart();
  }//reset the timer before we start
  
    if(spinTimer.hasPassed(spin_v)){
        spinTimer.restart();
        spinTimer.stop();
        spinDown();
        resultcode = 1;
 //       Serial.println("Spin Halted");
    }else{
        spinUp();
        resultcode = 0;
 //       Serial.println("Spin Started");     
    }
}
  digitalWrite(LedPin, ledState);
  digitalWrite(SpinActuatorPin, SpinState);
  digitalWrite(CWPin, CWState);
return resultcode;
}

int ballanceLoad(){
  int resultcode = 0;
  SpinState = HIGH;
  //balance_v
  //balanceTimer

//if(fill_time < minFillTime_v){
//    balance_v = minEmptyTime_v / 2;
//    balance_v = maxEmptyTime_v / 2;
//}else{
//    balance_v = maxEmptyTime_v / 2;
//}
    
if(digitalRead(SpinSensorPin) == 0){

  if(!balanceTimer.isRunning()){
   balanceTimer.restart();
  }//reset the timer before we start

// Primary Ballance Loop
  if(balanceTimer.hasPassed(balance_v)){
        balanceTimer.restart();
        balanceTimer.stop();
        resultcode = 1;
//        Serial.println("Ballance Spin Started");
  }else{
        
    if(!balanceLoopTimer.isRunning()){
    balanceLoopTimer.restart();
  }//reset the timer before we start

if(spinEnable == 0){
  if(balanceLoopTimer.hasPassed(balance_loop_v1)){
        spinEnable = 1;
        balanceLoopTimer.restart();
        balanceLoopTimer.stop();
//        Serial.println("Spin Enabled");
  }else{
      spinDown();
//      Serial.println("Stop Spin");
//      Serial.println(balance_v);
  }
}

if(spinEnable == 1){

    if(!spinTimer.isRunning()){
    spinTimer.restart();
  }//reset the timer before we start
  
    if(spinTimer.hasPassed(balance_loop_v2)){
        spinTimer.restart();
        spinTimer.stop();
        spinDown();
        spinEnable = 0;
//        Serial.println("Spin Dissabled");
    }else{
        spinUp();
//        Serial.println("Spinning");
//        Serial.println(balance_v);
    }
 }
} //Primary Ballance Loop

}//check the Spin Sensor      
  digitalWrite(LedPin, ledState);
  digitalWrite(SpinActuatorPin, SpinState);
  digitalWrite(CWPin, CWState);
  return resultcode; 
}

int rinseSpin(){
  int resultcode = 0;
  SpinState = HIGH;
  //ColdState = HIGH;

//if(fill_time < minFillTime_v){
//    balance_v = minEmptyTime_v / 2;
//    rinseSpin_v = maxEmptyTime_v;
//}else{
//    rinseSpin_v = maxEmptyTime_v;
//}

if(digitalRead(SpinSensorPin) == 0){
  if(!rinseSpinSpin.isRunning()){
    rinseSpinSpin.restart();
  }//reset the timer before we start

    if(rinseSpinSpin.hasPassed(rinseSpin_v)){
        rinseSpinSpin.restart();
        rinseSpinSpin.stop();
        spinDown();
        resultcode = 1;
 //       Serial.println("Rinnse Spin Halted");
    }else{
        spinUp();    
        resultcode = 0;
 //       Serial.println("Rinse Spin Started");     
    }    
}

  digitalWrite(LedPin, ledState);
  digitalWrite(SpinActuatorPin, SpinState);
  digitalWrite(CWPin, CWState);
//  digitalWrite(ColdWaterValve, ColdState);
return resultcode;
}

void coastSpin(){
  stopSpin();
  digitalWrite(LedPin, ledState);
//  digitalWrite(SpinActuatorPin, SpinState);
  digitalWrite(CWPin, CWState);
}

void endSpinCycle(){
  stopSpin();
  digitalWrite(LedPin, ledState);
  digitalWrite(SpinActuatorPin, SpinState);
  digitalWrite(CWPin, CWState);
}

int estop(){
  digitalWrite(ColdWaterValve, LOW);
  digitalWrite(HotWaterValve, LOW);
  digitalWrite(CWPin, LOW);
  digitalWrite(CCWPin, LOW);
  digitalWrite(SpinActuatorPin, LOW);
  digitalWrite(WaterPumpPin, LOW);

//Reset all timer's
fillTimer.restart();
fillTimer.stop();
agitateTimer.restart();
agitateTimer.stop();
agitateLoopTimer.restart();
agitateLoopTimer.stop();
balanceTimer.restart();
balanceTimer.stop();
balanceLoopTimer.restart();
balanceLoopTimer.stop();
spinTimer.restart();
spinTimer.stop();
rinseSpinSpin.restart();
rinseSpinSpin.stop();
rinseSpinWater.restart();
rinseSpinWater.stop();
waterFillTime.restart();
waterFillTime.stop();
waterEmptyTime.restart();
waterEmptyTime.stop();
soakTime.restart();
soakTime.stop();
spinBrakeTime.restart();
spinBrakeTime.stop();


//  Serial.println("WARNING: An error has occured and the system has halted");
//  Serial.print("Error Code : ");
//  Serial.println(error);

//while(RestartStatus != 0){
  //Loop and do noting util error is cleared but check pin
//  RestartStatus = digitalRead(RestartPin);
//  if(RestartStatus == 0){
//    Serial.print("Error Code : ");
//    Serial.print(error);
//    Serial.println(" Had been cleared and system is restarting");
//  }
//}
  MachineProcess = 0;
  return 0;
}

void pumpOn(){
   if(!waterEmptyTime.isRunning()){
//      Serial.println("Restarting Timer");
    waterEmptyTime.restart();
    }//reset the timer before we start

    PumpState = HIGH;   //Turn the pump on
    empty_time = waterEmptyTime.elapsed();
    digitalWrite(WaterPumpPin, PumpState);
}

void pumpOff(){
    PumpState = LOW;   //Turn the pump on
    waterEmptyTime.stop();
    digitalWrite(WaterPumpPin, PumpState);
//    Serial.print("Time taken to empty : ");
//    Serial.print(empty_time);
//    MachineProcess = 0;
}

int emptyMachine(){
    int resultcode = 0;
    int emptyTime = 0;
   //if the timer is not running then reset it
     if(!waterEmptyTime.isRunning()){
//      Serial.println("Restarting Timer");
      waterEmptyTime.restart();
    }//reset the timer before we start

    empty_time = waterEmptyTime.elapsed();

if(fill_time < 30){
//  fill_time = maxFillTime_v / 4;
    emptyTime = maxEmptyTime_v; 
}else{
    emptyTime = minEmptyTime_v;
}
    
    if(waterEmptyTime.hasPassed(emptyTime)){
      waterEmptyTime.stop();          
//      Serial.println("The bowl should be empty");
      PumpState = LOW;   //Turn the pump on
      MachineProcess = 0;
      resultcode = 1;
    } else {
//      Serial.println("Empty Bowl");
      PumpState = HIGH;   //Turn the pump on
      resultcode = 0;
    }
    digitalWrite(WaterPumpPin, PumpState); 
return resultcode;
}

void ColdWaterOn(){
  digitalWrite(LedPin, HIGH); //Start Filing
  digitalWrite(ColdWaterValve, HIGH);
}

void ColdWaterOff(){
  waterFillTime.restart();
  waterFillTime.stop();
  digitalWrite(LedPin, LOW); //Stop Filing
  digitalWrite(ColdWaterValve, LOW);
//  MachineProcess = 0;
}

void HotWaterOn(){
  digitalWrite(LedPin, HIGH); //Start Filing
  digitalWrite(HotWaterValve, HIGH);
}

void HotWaterOff(){
  digitalWrite(LedPin, LOW); //Stop Filing
  digitalWrite(HotWaterValve, LOW);
//  MachineProcess = 0;
}

int intSpray(){
      if(!rinseSpinWater.isRunning()){
          rinseSpinWater.restart();
      }//reset the timer before we start
  
      if(rinseSpinWater.hasPassed(rinseSpinWater_v)){
        rinseSpinWater.restart();
        rinseSpinWater.stop();
        if(ColdState == HIGH){
            ColdState = LOW;
        }else{
            ColdState = HIGH;
        }        
    }
    
digitalWrite(ColdWaterValve, ColdState);
//Serial.print("Water Valve: ");
//Serial.println(ColdState);
}

int fillWater(){
  int resultcode = 0;                           //clear and create the result code
  int WaterLevel = digitalRead(WaterLevelPin);  // Read the water level sensor
//  Serial.print("Water Level : ");
//  Serial.println(WaterLevel);
  
  if(!waterFillTime.isRunning()){
//    Serial.println("Restarting Timer");
    waterFillTime.restart();
  }//reset the timer before we start

  fill_time = waterFillTime.elapsed();          //See how many seconds it take to fill the machine
  
  if ((WaterLevel == 1) || (maxFillTime_v == fill_time)){
    digitalWrite(LedPin, LOW);                    //Stop Filling
    digitalWrite(ColdWaterValve, LOW);            // Turn off the cold water valve
    waterFillTime.stop();                         //Turn off the timer for the water fill
    resultcode = 1;                               //Filled with Water
    Serial.println("Machine Filled with Water");
  }//fill cycle conpleate

  if ((WaterLevel == 0) && (fill_time < maxFillTime_v)){
    digitalWrite(LedPin, HIGH);                   //Start Filing
    digitalWrite(ColdWaterValve, HIGH);           //turn on the cold water valve
    resultcode = 0;                             //Still filling with water
  }// filling washing machine
 return resultcode;
}
`

**Comment by vangalvin** [comment]
Great News, Added a snubber to each of the relay's that were handling the motors (240V) and it seems to have resolved the issue. 

[http://www.austech.info/showthread.php/31533-How-to-overcome-relay-contact-burn-out](http://www.austech.info/showthread.php/31533-How-to-overcome-relay-contact-burn-out)

**Comment by sofian** [comment]
Be careful with the comparison:
```maxFillTime_v == fill_time```

Because this will only be true at the exact second where the timer reaches the fill time. If for any reason you miss it, you'll never be able to react. It is error-prone.
Replace it with:
```maxFillTime_v <= fill_time```

Or just use the built-in function of Chrono:
```waterFillTime.hasPassed(maxFillTime_v)```

**Comment by vangalvin** [comment]
thanks @sofian  
I will make the suggested alteration. 

one of the things I am hoping to do is to capture the length of time it takes to fill the machine then map that with the maximum and minimum times it takes to empty the machine. sadly there is no way to check if the machine is empty and the mechanical method used for emptying is just set to how much time it takes to empty the machine when it full.

You can adjust the amount of water that goes in so a fill can take anywhere from 357 seconds to 539 seconds. the time it takes to pump the water out is 213 seconds to 300 seconds.
so the way I see it is i can reduce the time it takes for each load by doing something like...
map(fill_time, 357, 539, 213, 300) 

then I can split that time up for the empty and spin cycles and feed that in to Chrono. :)

the only issue i found was that map does not handle long's so the first time i tried i ended up with a negative number being returned. 

Once I am happy with the code I am going to post it on https://hackaday.io/project/20224-washing-machine-conversion and give Chrono a mention as it is prety integral to the operation of the final system.

B

**Comment by sofian** [comment]
Also, as I said earlier: the timers used in Chrono are based on millis(), which is not a perfect measure of time. The output might vary from time to time, and depending on the voltage and current fed to the circuit. If accuracy is important to you you have to use a RTC (like this one: https://www.sparkfun.com/products/12708)

**Comment by vangalvin** [comment]
Thanks. :) acuracey is not port and as things like water pressure change the fill time as we'll as the soap used can change the empty time. A +- 10 seconds is not going to be a big deal. I have added about 20 seconds to the empty time for this reason.

One day maybe adding a flow sensor to both full and empty would sort this bit until then keeping it kind of simple :)


Sent from my iPhone

On 9/03/2017, at 6:14 am, "Sofian Audry" <notifications@github.com<mailto:notifications@github.com>> wrote:


Also, as I said earlier: the timers used in Chrono are based on millis(), which is not a perfect measure of time. The output might vary from time to time, and depending on the voltage and current fed to the circuit. If accuracy is important to you you have to use a RTC (like this one: https://www.sparkfun.com/products/12708)

**Comment by vangalvin** [comment]
@sofian the maxFillTime_v == fill_time is actually only there to make sure that if the maximum fill time is exceeded the water valve shuts off. just in case for some reason the water level sensor fails to trigger it will not continue to fill and flood the floor :)
this is also why I used the N/O side of the sensor so if the sensor fails or a wire disconnects the fill will stop.
